src/appleseed/renderer/kernel/shading/shadingengine.cpp

```diff
@@ -97,25 +98,52 @@ void ShadingEngine::shade_hit_point(
     const ShadingContext&       shading_context,
     const ShadingPoint&         shading_point,
     AOVAccumulatorContainer&    aov_accumulators,
-    ShadingResult&              shading_result) const
+    ShadingResult&              shading_result,
+    Color4f&                    matte_color) const
 {
     // Compute the alpha channel of the main output.
     shading_result.m_main.a = shading_point.get_alpha()[0];
 
     // Retrieve the material of the intersected surface.
     const Material* material = shading_point.get_material();
 
-    // Apply OSL transparency if needed.
-    if (material &&
-        material->get_render_data().m_shader_group &&
-        material->get_render_data().m_shader_group->has_transparency())
+    if (material != nullptr && material->get_render_data().m_shader_group)
     {
-        Alpha alpha;
-        shading_context.execute_osl_transparency(
-            *material->get_render_data().m_shader_group,
-            shading_point,
-            alpha);
-        shading_result.m_main.a *= alpha[0];
+        const ShaderGroup* sg = material->get_render_data().m_shader_group;
+
+        if (sg->has_transparency() || sg->has_matte())
+        {
+            // This is similar to other functions but
+            // does not call process_xxx_tree().
+            // Folding the two cases together, we save one shading group exec.
+            shading_context.execute_osl_transparency_and_matte(
+                *material->get_render_data().m_shader_group,
+                shading_point);
+
+            if (sg->has_matte())
+            {
+                Color3f matte_rgb;
+                Alpha matte_alpha;
+                const bool any_matte_closure = process_matte_tree(
+                    shading_point.get_osl_shader_globals().Ci,
+                    matte_rgb,
+                    matte_alpha);
+
+                if (any_matte_closure)
+                {
+                    matte_color.rgb() = matte_rgb;
```

Comment:
can we write matte_color directly to shading_result?

Reply:
I moved matte check to sample renderer to break the loop somehow. Otherwise it renders everything beneath the matte surface as if it's not visible.

Reply:
I see. Maybe we can set matte color and matte alpha here and return any_matte_closure?
In the genericsamplerenderer if shade returns true, we can stop.